include/pybind11/iostream.h

```diff
@@ -68,7 +73,7 @@ class pythonbuf : public std::streambuf {
 
     /// Sync before destroy
     ~pythonbuf() override {
-        sync();
+        _sync();
```

Comment:
Independently of @bstaletic's experiments: why this way, and not just `pythonbuf::sync();`. That should also make it a non-virtual call, no?
I think the problem [your useful link](https://wiki.sei.cmu.edu/confluence/display/cplusplus/OOP50-CPP.+Do+not+invoke+virtual+functions+from+constructors+or+destructors) tries to solve with a private method is meant for if the `virtual` function would call through to the base class?

Reply:
Because `pythonbuf::sync()` didn't fix it. I haven't proven that this fixes it either, though.

Reply:
Huh? That's weird, though, no? If you take away the `virtual` nature of the call, if should be equivalent, or not?

Reply:
I'm not sure this fixes it, but is fills in so many points; the observed bug on MSVC would exactly match what this warning claims happens, and compilers like GCC optimize this by replacing the virtual call - thereby causing the correct behavior. Maybe it's not it, but I think we should keep this as a possible fix; and if we get the IO error again in the future, we can convert it to the fully qualified call (which also silences the warning). I think MSVC _might_ still treat this as an overloaded virtual function, even though you are trying to qualify it.

Reply:
> I think MSVC _might_ still treat this as an overloaded virtual function, even though you are trying to qualify it.

:roll_eyes:

But OK, I like this proposed approach! If it does fix things, we need to have it; otherwise, we can simplify :-)
You might want to add an extra line of comment underneath `// This function must be non-virtual to be called in a destructor` for future readers that this can't (currently) be simplified to the fully-qualified version, then?

Reply:
Done.

Reply:
Thanks!